app/src/main/java/com/github/orkest/View/sharing/SharingActivity.kt

```diff
@@ -63,35 +88,124 @@ class SharingComposeActivity : ComponentActivity() {
         }
     }
 
+    /**
+     * This function is used to connect to the Spotify API and get the song name
+     *
+     */
+    private fun spotifyAuthorization(){
+        // This is a callback for the authorization function that requests the access token
+        val showLoginActivityToken = registerForActivityResult(
+            ActivityResultContracts.StartActivityForResult()
+        ) { result: ActivityResult ->
+
+            val authorizationResponse = AuthorizationClient.getResponse(result.resultCode, result.data)
+
+            // Check the response type
+            when (authorizationResponse.type) {
+                AuthorizationResponse.Type.TOKEN -> {
+                    // Retrieve the access token
+                    this.accessToken = authorizationResponse.accessToken
+                    Log.d("SPOTIFY TOKEN", this.accessToken)
+                    CompletableFuture.runAsync {
+                        getTrackName(this.spotifySongID)
+                    }
+                }
+                AuthorizationResponse.Type.ERROR ->
+                    Log.d("SPOTIFY ERROR", authorizationResponse.error)
+                else ->
+                    Log.d("SPOTIFY FATAL ERROR", "Abort")
+            }
+        }
+
+        // This is a callback for the authorization function that requests the authorization code
+        val showLoginActivityCode = registerForActivityResult(
+            ActivityResultContracts.StartActivityForResult()
+        ) { result: ActivityResult ->
+
+            val authorizationResponse = AuthorizationClient.getResponse(result.resultCode, result.data)
+            when (authorizationResponse.type) {
+                AuthorizationResponse.Type.CODE -> {
+                    // retrieve the authorization code
+                    this.authorizationCode = authorizationResponse.code
+                    Log.d("SPOTIFY", "Code: ${this.authorizationCode}")
+                    CompletableFuture.runAsync {
+                        showLoginActivityToken.launch(getLoginActivityTokenIntent(authorizationCode, this))
+                    }
+                }
+                AuthorizationResponse.Type.ERROR ->
+                    // log the error
+                    Log.d("Debug", "Error: ${authorizationResponse.error}")
+                else ->
+                    // interrupt
+                    Log.d("Debug", "Error: Fatal")
+            }
+        }
+
+        val completableFutureSong = CompletableFuture.supplyAsync {
+            showLoginActivityCode.launch(requestUserAuthorization(this))
+        }
+
+        // retrieve the song name from the future
+        completableFutureSong.get()
```

Comment:
Make use of the .whenComplete() function of a future to make the code run asynchronously instead of synchronously

Reply:
It does run asynchronously. Everything inside the completableFuture trace is launched with runAsync so no problem with get(). Will double check if this is a good practice though.

Reply:
Indeed, the get() function always waits for the result, even if launched with runAsync before.